File: qualitystats.py
#!/usr/bin/python

__author__ = "Peter Williams"
__copyright__ = "Copyright 2016"
__credits__ = ["Peter Williams"]
__license__ = "GPL"
__version__ = "1.0.0"
__maintainer__ = "Peter Williams"
__email__ = "peter.williams@yale.edu"

### Usage:      python3 qualitystats.py -i <input file> -o <output file>
### Examples:   python3 qualitystats.py -i sample-input.fastq -o sample-output.txt
###				python3 qualitystats.py -i sample-input.fastq
### Note:       	Generates a set of 3 plots and a text file for corresponding fastq file
###					The -o flag is optional
###
### Input Formats:	-i 	string of corresponding fastq file
###			-o	string containing the name of the file to which the output information is saved
### Output Format:	txt file containing the file name, the number of sequences and the titles of corresponding plots

### Import libraries
import argparse
import numpy as np
import matplotlib.pyplot as mp
#import matplotlib as mpl
#rc_defaults=dict(mpl.rcParams)

### This is one way to read in arguments in Python. We need to read input file and score file.
parser = argparse.ArgumentParser(description='Quality Stats Generator')
parser.add_argument('-i', '--input', help='fastq file name', required=True)
parser.add_argument('-o', '--output', nargs='?',default='output.txt', help='Output file name')
args = parser.parse_args()

### Implementation
def qstats(filename,outfile):
	
	############################################################################
	### Create an array of q-scores
	
	### Initialize Variables and load fastq file
	qscores=np.zeros((1,1))
	count=0
	lengths=[]
	f=open(filename,'r')
	### Input quality scores, convert from ascii format, and assign to numpy array
	# 		If length of the reads is not the same, allow for variable length by
	#		making the array of dimension (number of reads)x(length of longest read).
	#		empty entries in the array will be assigned NAN so that stats will not be affected.
	for x in f:
		# If first score line, resize array, assign phred score converted value to last row
		if count==3:
			qscores.resize(1,len(x.strip()))
			qscores[len(qscores[:,0])-1,:]=np.array([int(ord(i)) for i in x.strip()])-33
			lengths.append(len(x.strip()))
		# If score line, resize array, assign phred score converted value to last row
		if count%4==3 and count>3:
			lengths.append(len(x.strip()))
			if len(x.strip())>len(qscores[0,:]):
				maxlen0=len(qscores[0,:])
				maxlen=len(x.strip())
				switch=1
			else:
				switch=0
			qscores.resize(len(qscores[:,0])+1,max(len(x.strip()),len(qscores[0,:])))
			if switch==1:
				qscores[:,maxlen0:maxlen]=np.NAN
			qscores[-1,:]=np.NAN
			qscores[len(qscores[:,0])-1,:]=np.array([int(ord(i)) for i in x.strip()])-33
		count+=1
	f.close()
	
	###########################################################################
	### Make Plots
	f=mp.figure(figsize=(5.2,4.5))
	mp.hist(np.array(lengths))
	mp.ylabel('Number of reads at given ')
	mp.title('Distribution of Sequence Lengths')
	mp.xlabel('Sequence Length (bp)')
	f.tight_layout()
	f.savefig('Sequence_Length_Distribution.png')

	f=mp.figure(figsize=(5.2,4.5))
	mp.errorbar(range(0,101),np.nanmean(qscores,axis=0),yerr=[np.nanstd(qscores,axis=0),np.nanstd(qscores,axis=0)])
	mp.errorbar(range(0,101),np.nanmedian(qscores,axis=0),yerr=[np.nanmedian(qscores,axis=0)-np.nanpercentile(qscores,25,axis=0),-np.nanmedian(qscores,axis=0)+np.nanpercentile(qscores,75,axis=0)])
	mp.xlabel('Position in Read (bp)')
	mp.ylabel('Qaulity Phred Score')
	mp.legend(['Mean and Standard Deviation','Median and Quartiles'],loc=0)
	f.tight_layout()
	f.savefig('Per_Base_Sequence_Quality.png')

	f=mp.figure(figsize=(5.2,4.5))
	mp.hist(np.nanmean(qscores,axis=1))
	mp.ylabel('Frequency of Mean Phred Value')
	mp.title('Sequence Quality Score Distribution')
	mp.xlabel('Mean Phred Score per Sequence')
	f.tight_layout()
	f.savefig('Per_Sequence_Mean_Quality_Distribution.png')
	
	###########################################################################
	### Create output file
	f=open(outfile,'w')
	f.write('Quality Score Statistics and Figure names for %s' %filename)
	f.write('\n\n')
	f.write('Number of Sequences = %d' %len(qscores[:,0]))
	f.write('\n\n')
	f.write('For the Distribution of read lengths see: Sequence_Length_Distribution.png')
	f.write('\n\n')
	f.write('For a plot of the per read quality see: Per_Base_Sequence_Quality.png')
	f.write('\n\n')
	f.write('For a plot of the distribution of mean quality per sequence see: Per_Sequence_Mean_Quality_Distribution.png')
	f.close()
	
### Run
qstats(args.input, args.output)

File: qualitystats.R
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Title :  qualitystats.R 
# Version : 1.0
#
# Purpose : A tool that outputs standard QC metrics and figures for fastq files
#  
# Version Notes : 
#
# Created.date  : 27 Apr 2016
# Created.by    : Dan Spakowicz
# Updated.date  :  
# Updated.by    : 
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

### Usage:      Rscript qualitystats.R -i <input file> -o <output file>
### Examples:   Rscript qualitystats.R -i sample-input.fastq -o sample-output.txt
###				Rscript qualitystats.R -i sample-input.fastq
### Note:       	Generates a set of 3 plots and a text file for corresponding fastq file
###					The -o flag is optional
###
### Input Formats:	-i 	string of corresponding fastq file
###			-o	string containing the name of the file to which the output information is saved
### Output Format:	txt file containing the file name, the number of sequences and the titles of corresponding plotsgenes

# remove any previously loaded functions
rm(list=ls())

# suppress warnings
oldw <- getOption("warn")
options(warn = -1)

# Load the required packages
list.of.packages <- c("optparse", "reshape2", "ggplot2")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)
library(optparse)
library(reshape2)
library(ggplot2)

if(!require(ShortRead)){
  source("http://bioconductor.org/biocLite.R")
  biocLite("ShortRead")
  library(ShortRead)
}
library(ShortRead)


# set arguments
option_list = list(
  make_option(c("-i", "--input"), type="character", default=NULL, 
              help="fastq file name", metavar="character"),
  make_option(c("-o", "--out"), type="character", 
              default="output_R", 
              help="output file name [default= %default]", metavar="character")
); 

opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);

#######
# create function for generating quality assessment plots
#######

QualityStats <- function(infile, outfile) {
  
  # input and process using shortRead
  input <- readFastq(infile)
  
  # scores for each nucleotide
  scores <- data.frame(as(quality(input), "matrix"))
  
  # lengths of each read
  lengths <- data.frame(apply(scores, 1, function(x) length(which(!is.na(x)))))
  
  #average quality score per nucleotide for each read
  ave.score <- (alphabetScore(input))/lengths
  
  #plot a histogram of the average quality score of each read
  ggplot(ave.score, aes(ave.score))+
    geom_histogram() +
    labs(x = "Quality Score",
         y = "Number of Reads",
         title = "Average Quality Scores for each Read")
  ggsave(paste(outfile, "_qual_hist.png", sep = ""))

  
  # plot a histogram of the average length of each read
  ggplot(lengths, aes(lengths))+
    geom_histogram()+
    labs(x = "Length",
         y = "Number of Reads",
         title = "Read Lengths")
  ggsave(paste(outfile, "_length_hist.png", sep = ""))
  
  
  # convert to long format for plotting boxplots by position
  m.scores <- melt(scores, na.rm = T)
  
  # calculate the means for each position  
  means <- data.frame(colMeans(scores, na.rm = T))
  means$id <- row.names(means)
  colnames(means) <- c("means", "id")
  
  # join the means with the long format data
  m.scores.means <- merge(m.scores, means, by.x = "variable", by.y = "id", all =T)
  colnames(m.scores.means) <- c("Position", "value", "means")
  m.scores.means$Position <- gsub("X", "", m.scores.means$Position)
  m.scores.means$Position <- as.integer(m.scores.means$Position)
  m.scores.means <- m.scores.means[order(m.scores.means$Position),]
  
  # plot boxplots of read qualities by position
  ggplot(m.scores.means, aes(x = factor(Position), y = value)) +
    geom_boxplot(fill = "green", outlier.size = 0) +
    geom_line(aes(x = Position, y = means, color = "Mean"), size = 1) +
    labs(x = "Position", y = "Quality Score", 
         title = "Read Quality Scores for each Position")+
    theme(axis.text.x = element_text(angle = 45),
          legend.position = c(0.9, 0.2), 
          legend.background = element_rect(color = "black", size = 1, linetype = "solid"))+
    scale_x_discrete(breaks = c(seq(1, nrow(means), by = 4)))+
    scale_color_manual("", values=c(Mean="red"))
  ggsave(paste(outfile, "_perBase_qual.png", sep = ""))
  
  
  # write summary text file 
  fileOut<-file(paste(outfile, ".txt", sep=""))
  writeLines(c(paste("Quality Score Statistics and Figure names for ", infile, sep=""),
               "",
               paste("Number of Sequences = ", nrow(scores), sep=""),
               "",
               paste("For the Distribution of read lengths see: ", outfile, "_length_hist.png", sep = ""),
               "",
               paste("For a plot of the per base quality score see: ", outfile, "_perBase_qual.png", sep = ""),
               "",
               paste("For a plot of the distribution of mean quality per sequence see: ", outfile, "_qual_hist.png", sep = "")),
             fileOut)
  close(fileOut)
}

QualityStats(opt$input, opt$out)

# return warning setting
options(warn = oldw)